src/themes/default/pages/Product.vue

```diff
@@ -260,85 +266,84 @@ export default {
-          const type = configurableOption.attribute_code
-          const filterVariants = configurableOption.values.map(
-            ({ value_index, label }) => {
-              let currentVariant = this.options[type].find(
-                config => config.id === value_index
-              )
-              label =
-                label || (currentVariant ? currentVariant.label : value_index)
-              return { id: value_index, label, type }
-            }
-          )
-          const availableOptions = filterVariants.filter(option =>
-            this.isOptionAvailable(option)
-          )
-          filtersMap[type] = availableOptions
-        })
+    offlineImage () {
+      return {
+        src: this.getThumbnail(this.getCurrentProduct.image, config.products.thumbnails.width, config.products.thumbnails.height),
+        error: this.getThumbnail(this.getCurrentProduct.image, config.products.thumbnails.width, config.products.thumbnails.height),
+        loading: this.getThumbnail(this.getCurrentProduct.image, config.products.thumbnails.width, config.products.thumbnails.height)
       }
-      return filtersMap
+    },
+    customAttributes () {
+      return Object.values(this.attributesByCode).filter(a => {
+        return a.is_visible && a.is_user_defined && (parseInt(a.is_visible_on_front) || a.is_visible_on_front === true) && this.getCurrentProduct[a.attribute_code]
+      })
+    },
+    getAvailableFilters () {
+      return getAvailableFiltersByProduct(this.getCurrentProduct)
     },
     getSelectedFilters () {
-      // TODO move to helper when refactoring product page
-      let selectedFilters = {}
-      if (this.configuration && this.product) {
-        Object.keys(this.configuration).map(filterType => {
-          const filter = this.configuration[filterType]
-          selectedFilters[filterType] = {
-            id: filter.id,
-            label: filter.label,
-            type: filterType
-          }
-        })
-      }
-      return selectedFilters
+      return getSelectedFiltersByProduct(this.getCurrentProduct, this.getCurrentProductConfiguration)
     },
     isSimpleOrConfigurable () {
       if (
-        this.product.type_id === 'simple' ||
-        this.product.type_id === 'configurable'
+        this.getCurrentProduct.type_id === 'simple' ||
+        this.getCurrentProduct.type_id === 'configurable'
       ) { return true }
       return false
     },
     getInputName () {
-      if (this.isSimpleOrConfigurable) { return i18n.t('Quantity available', { qty: this.quantity }) }
-      return i18n.t('Quantity')
+      if (this.isSimpleOrConfigurable && !this.isProductLoading) { return this.$i18n.t('Quantity available', { qty: this.quantity }) }
+      return this.$i18n.t('Quantity')
     }
   },
   created () {
     this.getQuantity()
   },
+  mounted () {
+    this.$store.dispatch('recently-viewed/addItem', this.getCurrentProduct)
+  },
+  async asyncData ({ store, route }) {
```

Comment:
Please make sure it works well with navigating thru related products (so on the same `Product` component - just switching the router parameters; I'd encountered a lot of issues with this behavior when working on url dispatcher0